File: java/CombinationSumIII.java
/*
 * Filename: CombinationSumIII.java
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Mon 13 Jul 2015 01:53:07 PM EDT
 */

public class Solution {
	public List<List<Integer>> combinationSum3(int k, int n) {
		List<List<Integer>> res = new ArrayList<List<Integer>>();
		List<Integer> list = new ArrayList<Integer>();
		combinationSum3Aux(k, n, res, list, 0);
		return res;
	}

	private void combinationSum3Aux
			(int k, int n, List<List<Integer>> res, List<Integer> list, int i) {
		if (n < 0) {
			return;
		}

		if (i == k) {
			if (n == 0)	{
				List<Integer> list2 = new ArrayList<Integer>();
				list2.addAll(list);
				res.add(list2);
			}

			return;
		}

		int stt = 1;
		if (list.size() != 0) {
			stt += list.get(list.size() - 1);
		}

		for (int j = stt; j <= 9; j++) {
			list.add(j);
			combinationSum3Aux(k, n-j, res, list, i+1);
			list.remove(list.size() - 1);
		}
	}
}

File: cpp/combination_sum_iii.cc

/*
 * Filename: combination_sum_iii.cc
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Fri 14 Aug 2015 03:31:40 PM EDT
 */

#include <iostream>
using namespace std;

class Solution {
	public:
		vector<vector<int>> combinationSum3(int k, int n) {
			vector<vector<int> > res;
			vector<int> tuple;
			combination_sum3_aux(0, k, n, res, tuple);
			return res;
		}

	private:
		void combination_sum3_aux(
				int i,
				int k,
				int n,
				vector<vector<int> >& res,
				vector<int>& tuple)
		{
			if (k < 0 || n < 0)
			{
				return;
			}

			if (!k && !n)
			{
				res.push_back(tuple);
				return;
			}

			for (int j = i+1; j <= 9; j++)
			{
				tuple.push_back(j);
				combination_sum3_aux(j, k-1, n-j, res, tuple);
				tuple.pop_back();
			}
		}
};
